# HEROES of PYMOLI
![Heroes Logo](resources/fantasy.jpg)
### Game Analysis

In [2]:
#import pandas

import pandas as pd

In [3]:
#Import Modules
import numpy as np
import os
import csv

In [4]:
#Set path for file import

csvpath = os.path.join("Resources", "purchase_data.csv")

In [5]:
#Set table to data frame

df_purchase_data = pd.read_csv(csvpath)
df_purchase_data.head()

Purchase_ID             SN  Age Gender  Item_ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item_Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Number of Players
There are 576 unique individual players in the game.

In [6]:
#DETERMINE PLAYER COUNT

unique_players = pd.DataFrame([{"Individual Players": df_purchase_data["SN"].nunique()}])
unique_players


Individual Players
0                 576

In [7]:
#ANALYZE PURCHASING TOTALS:
#Number of Unique Items, Average Purchase Price, Total Number of Purchases, Total Revenue

## Purchasing Totals
There are 179 unique items purchased, with an average price of 3.05 each. 780 items were purchased, providing a total of 2379.77.

In [8]:
#Number of unique items
unique_items = pd.DataFrame([{"Item Name": df_purchase_data["Item_Name"].nunique()}])
unique_items

Item Name
0        179

In [9]:
#Average Purchase Price:
purchase_price = pd.DataFrame([{"Average Price": df_purchase_data["Price"].mean()}])
#Round average purchase price:
purchase_price.round(2)

Average Price
0           3.05

In [10]:
#Total Number of Purchases:
total_purchases = pd.DataFrame([{"Total Purchases": df_purchase_data["Price"].count()}])
total_purchases

Total Purchases
0              780

In [11]:
#Total Revenue
total_revenue = df_purchase_data["Price"].sum()
total_revenue

2379.77

In [12]:
#ANALYZE GENDER DEMOGRAPHICS:
# % and Count of Male Players, 
# % and Count of Female Players, 
# % and Count of Other/Non-disclosed Players

## Gender Demographics
Out of our total 576 players, identified males dominate the game as 484 of the players, or 84.03 percent. Females accounted for 81 of the players, or a mere 14.06 percent. Eleven players, or 1.91 percent, did not disclose a gender.

In [35]:
#Count of players by gender category, and percentage by gender

#Create dataframe that indicates the number of each gender category                  
Gender = df_purchase_data.groupby("Gender").nunique()
Male = Gender.iloc[1,1]
Female = Gender.iloc[0,1]
Other = Gender.iloc[2,1]

#Create a dataframe that calculates the totals and percentages of each gender category
df_percentage = pd.DataFrame({
    "Gender": ["Male", "Female", "Other/Not Disclosed"],
    "Total": [Male, Female, Other],
    "Percentage": [100*Male/df_purchase_data["SN"].nunique(), 
                    100*Female/df_purchase_data["SN"].nunique(), 
                    100*Other/df_purchase_data["SN"].nunique()]
})

df_percentage.loc["Total Players"]= df_percentage.sum(axis=0,numeric_only=True)
df_percentage.loc["Total Players", "Gender"] = ""

#Format the percentages to two decimals
df_percentage = df_percentage.round(2)

#Print out the total of each gender category and its percentage in a table
df_percentage

Gender  Total  Percentage
0                             Male  484.0       84.03
1                           Female   81.0       14.06
2              Other/Not Disclosed   11.0        1.91
Total Players                       576.0      100.00

In [14]:
#Analyze Purchasing by Gender: 

#For each gender, determine
#Purchase Count, Average Purchase Price, Total Purchase Value

## Purchasing by Gender
Because our game is male dominated, they accounted for the most purchases at 652. Even though their average purchase price was lower at 3.02 than the female and other/non-disclosed categories, they still had the highest purchase value, coming in at 1967.64. The other/nondisclosed category of players tended to spend more with an average of 3.35 per in-game purchase, but since this group made only 15 purchases total, they accounted for the lowest total purchase value at 50.19. Females also tended to spend more money than males, at 3.20 per in-game purchase, and with a total of 113 purchases, accounted for 361.94 of our total purchase value.

In [15]:
df_gender_2 = df_purchase_data.groupby('Gender')

Total_Purchase_Counts = round(df_gender_2['Price'].count(), 0)
Avg_Purchase_Price = round(df_gender_2['Price'].mean(), 2)
Total_Purchase_Values = round(df_gender_2['Price'].sum(), 2)
Purchase_Value_Per_Gender = round(Total_Purchase_Values / Total_Purchase_Counts, 2)

summary_dataframe2 = pd.DataFrame([Total_Purchase_Counts, Avg_Purchase_Price, Total_Purchase_Values, Purchase_Value_Per_Gender])
summary2 = summary_dataframe2.T
summary2.columns = ['Total Purchase Count', 'Average Purchase Price', 'Total Purchase Value', 'Avg Total Purchase per Person']
summary2

Total Purchase Count  Average Purchase Price  \
Gender                                                                
Female                                113.0                    3.20   
Male                                  652.0                    3.02   
Other / Non-Disclosed                  15.0                    3.35   

                       Total Purchase Value  Avg Total Purchase per Person  
Gender                                                                      
Female                               361.94                           3.20  
Male                                1967.64                           3.02  
Other / Non-Disclosed                 50.19                           3.35

In [16]:
#ANALYZE AGE DEMOGRAPHICS:

In [17]:
#* The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.)
#  * Purchase Count
#  * Average Purchase Price
#  * Total Purchase Value
#  * Average Purchase Total per Person by Age Group

## Age Demographics and Purchasing Analysis
Our players fell into 8 different age groups, with the largest age group being between 20 and 24, making up 258 of our players, or 44.79 percent. Ages 15-19 were our second most populous group, accounting for 18.58 percent of our players. The next largest groups fell between the ages of 25 and 39, making up 26.38 percent of our players. Those under 10 and over 40 were our least populated age groups at just over 2 percent of players each.

Surprisingly, the under 10 age group spent the most money per purchase, but because so many of our players were between the ages of 15-19 and 20-24, they spent 412.89 and 1114.06 in total in-game purchases, repspectively.

Except for the outliers of the under 10 crowd, it was a trend to see older players in the 35-39 age range spend more per purchase than the others, presumably because they have more money to spend. For players over 40, objectivity in spending money in-game appears to take hold, as their average purchases seemed to be lower.

In [18]:
#Determine age ranges broken into bins of 4 years
Age = df_purchase_data.groupby("SN").mean()
groups = [0, 9, 14, 19, 24, 29, 34, 39,  np.inf]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
Age["Bin"] = pd.cut(Age["Age"], groups, labels=group_names)
Bins = pd.DataFrame({"Total Count": Age.groupby("Bin")["Age"].count()})
Bins["Percentage of Players"] = 100* Bins["Total Count"]/Bins["Total Count"].sum()
Bins_formatted = Bins
Bins_formatted["Percentage of Players"] = Bins_formatted["Percentage of Players"].map("{:.2f}%".format)
Bins_formatted

Total Count Percentage of Players
Bin                                     
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

In [19]:
#For each age group (bin), determine:
#  * Purchase Count
#  * Average Purchase Price
#  * Total Purchase Value
#  * Average Purchase Total per Person by Age Group

df_purchase_data["Bin"] = pd.cut(df_purchase_data["Age"], groups, labels=group_names)
df_purchase_data_grouped = pd.DataFrame({
                    "Purchase Count": df_purchase_data.groupby("Bin")["Price"].count(),
                    "Average Purchase Price": round(df_purchase_data.groupby("Bin")["Price"].mean(),2),
                    "Total Purchase Value": 
                        df_purchase_data.groupby("Bin")["Price"].mean()*
                        df_purchase_data.groupby("Bin")["Price"].count(),
                    "Average Total per Person":
                        round(df_purchase_data.groupby("Bin")["Price"].mean()*
                        df_purchase_data.groupby("Bin")["Price"].count()/Bins["Total Count"],2)
                        })
df_purchase_data_grouped

Purchase Count  Average Purchase Price  Total Purchase Value  \
Bin                                                                   
<10                23                    3.35                 77.13   
10-14              28                    2.96                 82.78   
15-19             136                    3.04                412.89   
20-24             365                    3.05               1114.06   
25-29             101                    2.90                293.00   
30-34              73                    2.93                214.00   
35-39              41                    3.60                147.67   
40+                13                    2.94                 38.24   

       Average Total per Person  
Bin                              
<10                        4.54  
10-14                      3.76  
15-19                      3.86  
20-24                      4.32  
25-29                      3.81  
30-34                      4.12  
35-39                      4.76  
40+                        3.19

In [20]:
#ANALYZE TOP SPENDERS:

## Top Five Spenders
The top 5 spenders in the game are provided below by name. They are listed by the most money spent, which is differentiated by the average purchase price. For example, player name Lisosia93 only spent on average 3.79 per purchase, but since they made five purchases, their total purchase price was highest at 18.96. Similarly, player Iskadarya95 spent 4.37 on average per purchase, but only made 3 purchases, so their total purchase value was lowest at 13.10.

In [21]:
#Determine top 5 spenders by purchase price

df_description = df_purchase_data.groupby("SN")["Price"].describe()
df_top_spenders = pd.DataFrame({
    "Purchase Count": df_description["count"],
    "Average Purchase Price": df_description["mean"],
    "Total Purchase Value": df_description["mean"]*df_description["count"]
})
df_top_spenders = df_top_spenders.sort_values("Total Purchase Value", ascending=False)
df_top_spenders.head(5)

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93               5.0                3.792000                 18.96
Idastidru52             4.0                3.862500                 15.45
Chamjask73              3.0                4.610000                 13.83
Iral74                  4.0                3.405000                 13.62
Iskadarya95             3.0                4.366667                 13.10

In [22]:
#DETERMINE MOST POPULAR ITEMS

## Top Five Most Popular Items
The five most popular items that players were purchasing are listed below, and those that were bought the most were most often higher in total purchase value. The item with the name Final Critic was purchased 13 times at a price of 4.61, accounting for a total purchase price of 59.99. However, the item titled Fiery Glass Crusader, purchases 9 times and priced at 4.58, provided a purchase price of 41.22. In order to increase total purchase value, price of the item is probably better off staying lower. A lower price will be more appealing to players, and as they purchase more of an item, the total purchase price will increase.

In [23]:
#Determine top 5 most popular items by purchase count and generate table that describes
#these items with Item ID, Item Name, Purchase Count, Item Price, Total Purchase Value

Most_Popular_Items = df_purchase_data.drop(columns=["SN", "Age", "Gender", "Purchase_ID"])
Most_Popular_Items = Most_Popular_Items.groupby(["Item_ID", "Item_Name"])['Price'].describe()
Most_Popular_Items = Most_Popular_Items.drop(columns=["std", "min", "25%", "50%", "75%", "max"])
Most_Popular_Items["Total Purchase Value"] = Most_Popular_Items["count"]*Most_Popular_Items["mean"]
Most_Popular_Items = Most_Popular_Items.rename(columns={"Item_ID": "Item ID", "Item_Name": "Item Name", "count": "Purchase Count", "mean": "Item Price"})
Most_Popular_Items = Most_Popular_Items.sort_values("Purchase Count", ascending=False)
Most_Popular_Items.head()

Purchase Count  \
Item_ID Item_Name                                                      
92      Final Critic                                            13.0   
178     Oathbreaker, Last Hope of the Breaking Storm            12.0   
145     Fiery Glass Crusader                                     9.0   
132     Persuasion                                               9.0   
108     Extraction, Quickblade Of Trembling Hands                9.0   

                                                      Item Price  \
Item_ID Item_Name                                                  
92      Final Critic                                    4.614615   
178     Oathbreaker, Last Hope of the Breaking Storm    4.230000   
145     Fiery Glass Crusader                            4.580000   
132     Persuasion                                      3.221111   
108     Extraction, Quickblade Of Trembling Hands       3.530000   

                                                      Total Purchase Value  
Item_ID Item_Name                                                           
92      Final Critic                                                 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
145     Fiery Glass Crusader                                         41.22  
132     Persuasion                                                   28.99  
108     Extraction, Quickblade Of Trembling Hands                    31.77

In [24]:
#DETERMINE MOST PROFITABLE ITEMS

## Top Five Most Profitable Items
The five most profitable items list is similar to the most "popular" items list, simply because popular items are listed more often, and therefore make for more profit. Not all were the same, however, because higher prices that players are willing to pay increase the profitability of the item. Final Critic was the most purchased and highest priced item, so it was most profitable. However, an item like Nirvana can be priced higher, sell slightly less times, and also make the top five in profitability.

In [33]:
#Identify the 5 most profitable items by total purchase value
Most_Profitable_Items = Most_Popular_Items.sort_values("Total Purchase Value", ascending=False)

In [34]:
#List them in a table with the following information
#Item ID, Item Name, Purchase Count, Item Price, Total Purchase Value
Most_Profitable_Items.head(5)

Purchase Count  \
Item_ID Item_Name                                                      
92      Final Critic                                            13.0   
178     Oathbreaker, Last Hope of the Breaking Storm            12.0   
82      Nirvana                                                  9.0   
145     Fiery Glass Crusader                                     9.0   
103     Singed Scalpel                                           8.0   

                                                      Item Price  \
Item_ID Item_Name                                                  
92      Final Critic                                    4.614615   
178     Oathbreaker, Last Hope of the Breaking Storm    4.230000   
82      Nirvana                                         4.900000   
145     Fiery Glass Crusader                            4.580000   
103     Singed Scalpel                                  4.350000   

                                                      Total Purchase Value  
Item_ID Item_Name                                                           
92      Final Critic                                                 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
82      Nirvana                                                      44.10  
145     Fiery Glass Crusader                                         41.22  
103     Singed Scalpel                                               34.80